File: linked_list/linked_list/check_cll.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

def insert(head, data):
    ptr = Node(data)
    if head is None:
        ptr.next = ptr
        ptr.prev = ptr
        return ptr
    else:
        p = head
        while p.next != head:
            p = p.next
        p.next = ptr
        ptr.prev = p
        ptr.next = head
        head.prev = ptr
        return head

def check(head):
    if head is None:
        print("NULL")
        return True  # GFG convention: empty list is circular
    else:
        p = head
        while p.next is not None and p.next != head:
            p = p.next
        if p.next == head:
            return True
        else:
            return False

def display(head):
    if head is None:
        print("NULL")
        return
    p = head
    while p.next != head:
        print(p.data, end="->")
        p = p.next
    print(p.data)

if __name__ == "__main__":
    head = None
    head = insert(head, 10)
    head = insert(head, 10)
    head = insert(head, 30)
    t = check(head)
    print("Is circular?", "true" if t else "false")
    display(head)

File: linked_list/linked_list/check_cll.c
#include<stdio.h>
#include<stdlib.h>
#include<stdbool.h>
typedef struct Node{
    int data;
    struct Node *next;
    struct Node *prev;
} Node;
Node *insert(Node *head,int data){
    Node *p=head;
    Node *ptr=(Node *)calloc(1,sizeof(Node));
    ptr->data=data;
    if (head==NULL){
        ptr->next=ptr;
        ptr->prev=ptr;
        return ptr;
    }
    else{
        while(p->next!=head){
            p=p->next;
        }
        p->next=ptr;
        ptr->prev=p;
        ptr->next=head;
        head->prev=ptr;
        return head;
    }
}
bool check(Node *head){
    if (head==NULL){
        printf("NULL");
        return true;
    }
    else{
        Node *p=head;
        while(p->next!=head && p->next!=NULL){
            p=p->next;
        }
        if(p->next==head){
            return true;
        }
        else{
            return false;
        }
    }
}
void display(Node *head){
    Node *p=head;
    if(head==NULL){
        printf("NULL");
    }
    else{
        while(p->next!=head){
            printf("%d->",p->data);
            p=p->next;
        }
        printf("%d",p->data);
    }
}
int main(){
    Node *head=NULL;
    head=insert(head,10);
    head=insert(head,10);
    head=insert(head,30);
    bool t=check(head);
    printf("Is circular? %s\n", t ? "true" : "false");
    display(head);
    return 0;
}